var/spack/repos/builtin/packages/spark/package.py

```diff
@@ -14,14 +14,17 @@ class Spark(Package):
     """
 
     homepage = "http://spark.apache.org"
-    url = "http://archive.apache.org/dist/spark/spark-2.0.0/spark-2.0.0-bin-without-hadoop.tgz"
 
     variant('hadoop', default=False,
             description='Build with Hadoop')
 
-    depends_on('java', type=('build', 'run'))
-    depends_on('hadoop', when='+hadoop', type=('build', 'run'))
+    depends_on('java@8', type=('build', 'run'))
+    # Hadoop dependency after 2.1.0 will be included in downloaded tarball.
+    # ~hadoop + external Hadoop will be missing Hive et al support.
+    depends_on('hadoop', when='@:2.3.0 +hadoop', type=('build', 'run'))
 
+    version('2.4.0', 'c93c096c8d64062345b26b34c85127a6848cff95a4bb829333a06b83222a5cfa', when='+hadoop')
+    version('2.4.0', 'b1d6d6cb49d8253b36df8372a722292bb323bd16315d83f0b0bafb66a4154ef2', when='~hadoop')
```

Comment:
Is the `when=` in the `version` directive working at all? I couldn't find mention of it [here](https://spack.readthedocs.io/en/latest/packaging_guide.html#versions-and-fetching) and quickly peeking at the code I don't see where this is used. On top of that:
```console
$ spack fetch spark+hadoop
==> Fetching http://archive.apache.org/dist/spark/spark-2.4.0/spark-2.4.0-bin-hadoop2.7.tgz
################################################################################################################################################################################## 100,0%
==> Error: sha256 checksum failed for /home/mculpo/PycharmProjects/spack/var/spack/stage/spark-2.4.0-6aabr7avffnwxyrb75qnjz42airgol2z/spark-2.4.0-bin-hadoop2.7.tgz
Expected b1d6d6cb49d8253b36df8372a722292bb323bd16315d83f0b0bafb66a4154ef2 but got c93c096c8d64062345b26b34c85127a6848cff95a4bb829333a06b83222a5cfa
```
I think we are missing some support in the core of Spack to do what you propose here cleanly. @tgamblin @scheibelp @adamjstewart

Reply:
I would absolutely love to see the version directive get a `when` parameter. I've been waiting so long to be able to install `jdk` on macOS, but it's a different tarball.

Reply:
Also see #2009 for historical discussions of this topic.

Reply:
So… see the current status for a more hackish way. Of course, this breaks with source tarball caching.